tools/discovery/discovery.go

```diff
@@ -0,0 +1,131 @@
+package discovery
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+
+	"github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
+	"github.com/xiekeyang/oci-discovery/tools/object"
+)
+
+var (
+	defaultTrans = &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}
+)
+
+func DiscoveryHandler(context *cli.Context) error {
+	var (
+		name  = context.Args()[0]
+		roots []v1.Descriptor
+	)
+
+	v, err := paramsParser(name)
+	if err != nil {
+		return err
+	}
+
+	engines, err := refEnginesFetching(v)
+	if err != nil {
+		return err
+	}
+
+	for _, engine := range engines.RefEngines {
+		var ur urlResolver = urlResolver(engine.Uri)
+		u, err := ur.resolve(v)
+		if err != nil {
+			return err
+		}
+
+		index, err := ociIndexFetching(u)
+		if err != nil {
+			return err
+		}
+
+		if fragment, ok := v["fragment"]; ok {
+			for _, manifest := range index.Manifests {
+				if fragment == manifest.Annotations[`org.opencontainers.image.ref.name`] {
+					roots = append(roots, manifest)
+				}
+			}
+		} else {
+			roots = append(roots, index.Manifests...)
+		}
+	}
+
+	return stdWrite(roots)
+}
+
+func refEnginesFetching(v map[string]interface{}) (*object.RefEngines, error) {
+	u, err := templateRefEngines.resolve(v)
+	if err != nil {
+		return nil, err
+	}
+
+	client := &http.Client{Transport: defaultTrans}
+
+	resp, err := client.Get(u.String())
```

Comment:
This should be setting an `Accept` header for [`application/vnd.oci.ref-engines.v1+json`][1] and [checking that the response has that media type][2].

[1]: https://github.com/xiekeyang/oci-discovery/blob/44ec3cf3113e29a743ad04220ccb7ff5197dab2a/ref-engine-discovery.md#ref-engines-media-types
[2]: https://github.com/xiekeyang/oci-discovery/blob/44ec3cf3113e29a743ad04220ccb7ff5197dab2a/oci_discovery/fetch_json/__init__.py#L22-L25